client/remoterepo.c

```diff
@@ -106,11 +113,490 @@ _handle_curl_cb(
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    dwError = TDNFAllocateString(fileinfo->name, &(metalink_file->filename));
+    BAIL_ON_TDNF_ERROR(dwError); 
+    
+    metalink_file->digest = NULL;
+    if(fileinfo->checksums)
+    {
+        metalink_digest_alias *digest_alias;
+        for(digest_alias = digest_aliases; digest_alias->alias_name; ++digest_alias)
+        {
+            metalink_checksum_t **metalink_checksum;
+            for(metalink_checksum = fileinfo->checksums; *metalink_checksum; ++metalink_checksum)
+            {
+                if(!strcmp(digest_alias->alias_name, (*metalink_checksum)->type) &&
+                   TDNF_Check_Hex_Digest((*metalink_checksum)->hash, digest_alias->length))
+                {
+                    dwError = TDNFAllocateString((*metalink_checksum)->type, &(metalink_file->type));
+                    BAIL_ON_TDNF_ERROR(dwError);
+                    dwError = TDNF_New_Metalink_Checksum_From_Hex_Digest((*metalink_checksum)->hash,
+                                                                        &(metalink_file->digest));
+                    BAIL_ON_TDNF_ERROR(dwError);
+                    break;
+                }
+            }
+            if(metalink_file->digest)
+            {
+                break;
+            }
+        }
+    }
+    if(fileinfo->resources)
+    {
+        metalink_resource_t **metalink_resource = fileinfo->resources;
+        /* Filter by type if it is non-NULL. In Metalink v3, type
+           includes the type of the resource. In repo, we are only
+           interested in HTTP, HTTPS ,FTP and rsync.
+        */
+        if((*metalink_resource)->type == NULL ||
+           !strcmp((*metalink_resource)->type, "http") ||
+           !strcmp((*metalink_resource)->type, "https") ||
+           !strcmp((*metalink_resource)->type, "ftp") ||
+           !strcmp((*metalink_resource)->type, "ftps") ||
+           !strcmp((*metalink_resource)->type, "rsync"))
+        {
+            if((*metalink_resource)->url == NULL)
+            {
+                dwError = ERROR_TDNF_INVALID_REPO_FILE;
+                BAIL_ON_TDNF_ERROR(dwError); 
+            }
+            ***ml_file = metalink_file;
+        }
+        else
+        {
+            dwError = ERROR_TDNF_INVALID_REPO_FILE;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+    }
+
+cleanup:
+    return dwError;
+error:
+    if(metalink_file->filename)
+    {
+        TDNF_SAFE_FREE_MEMORY(metalink_file->filename);
+    }
+    if(metalink_file->type)
+    {
+        TDNF_SAFE_FREE_MEMORY(metalink_file->type);
+    }
+    if(metalink_file->digest)
+    {
+        TDNF_SAFE_FREE_MEMORY(metalink_file->digest);
+    }
+    if(metalink_file)
+    {
+        TDNF_SAFE_FREE_MEMORY(metalink_file);
+    }
+    goto cleanup;  
+}
 uint32_t
 TDNFParseAndGetURLFromMetalink(
     PTDNF pTdnf,
     const char *pszRepo,
-    const char *pszFile
+    const char *pszFile,
+    metalinkfile ***ml_file
```

Comment:
validate ml_file for null

Reply:
In case of just parsing metalink, we will not download anything.. Just parse the existing metalink to set the base url. In that case ml_file is expected to be null.

In this function, before calling  TDNF_New_Metalinkfile we are checking if ml_file is null or not. As the only usage for ml_file is in  TDNF_New_Metalinkfile.

Reply:
Okay, understood. But it'll be good if we have some kind of check to ensure that ml_file is non null when we are not parsing metalink file

Reply:
thats taken care.